File: src/me/xerces/eventbus/SimpleEventManager.java
package me.xerces.eventbus;

import me.xerces.eventbus.annotation.EventHandler;
import me.xerces.eventbus.events.Event;
import me.xerces.eventbus.annotation.EventPriority;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.*;

public class SimpleEventManager implements IEventManager{

    private Map<Class<?extends Event>, List<EventHandle>> eventHandleMap = new HashMap<>();

    /**
     * Add an object as an events listener, this method will loop through the objects methods
     * and if the method has {@link me.xerces.eventbus.annotation.EventHandler} present then it will add it has an Event Handler
     * @param object the object to register
     */
    public void addEventListener(Object object)
    {
        for(Method method : object.getClass().getDeclaredMethods())
        {
            if(method.isAnnotationPresent(EventHandler.class) && method.getParameterTypes().length > 0)
            {
                EventHandler eventHandler = (EventHandler)method.getDeclaredAnnotations()[0];
                if(eventHandleMap.get(method.getParameterTypes()[0]) != null)
                {
                    eventHandleMap.get(method.getParameterTypes()[0]).add(new EventHandle(method, object, eventHandler.priority()));
                    sortArrayList(eventHandleMap.get(method.getParameterTypes()[0]));
                } else {
                    List<EventHandle> eventHandles = new ArrayList<>();
                    eventHandles.add(new EventHandle(method, object, eventHandler.priority()));
                    eventHandleMap.put((Class<?extends Event>)method.getParameterTypes()[0], eventHandles);
                }
                method.setAccessible(true);
            }
        }
    }

    /**
     * Add an object as an events listener, this method will loop through the objects methods
     * and if the method has {@link me.xerces.eventbus.annotation.EventHandler} present then it will add it has an Event Handler
     * @param object the object to register
     * @param eventClass the specific events class we want to filter for
     */
    public void addSpecificEventListener(Object object, Class<?extends Event> eventClass)
    {
        for(Method method : object.getClass().getDeclaredMethods())
        {
            if(method.isAnnotationPresent(EventHandler.class) && method.getParameterTypes().length > 0 && method.getParameterTypes()[0].equals(eventClass))
            {
                EventHandler eventHandler = (EventHandler)method.getDeclaredAnnotations()[0];
                if(eventHandleMap.get(method.getParameterTypes()[0]) != null)
                {
                    eventHandleMap.get(method.getParameterTypes()[0]).add(new EventHandle(method, object, eventHandler.priority()));
                    sortArrayList(eventHandleMap.get(method.getParameterTypes()[0]));
                } else {
                    List<EventHandle> eventHandles = new ArrayList<>();
                    eventHandles.add(new EventHandle(method, object, eventHandler.priority()));
                    eventHandleMap.put((Class<?extends Event>)method.getParameterTypes()[0], eventHandles);
                }
                method.setAccessible(true);
            }
        }
    }

    /**
     * Sort the ArrayList by the Priority making sure its Highest - Lowest
     * @param arrayList the ArrayList to sort
     */
    private void sortArrayList(List<EventHandle> arrayList)
    {
        Collections.sort(arrayList, new Comparator() {
            @Override
            public int compare(Object objOne, Object objTwo) {
                EventHandle eventHandleFirst = (EventHandle) objOne;
                EventHandle eventHandleSecond = (EventHandle) objTwo;
                return eventHandleSecond.getPriority().ordinal() - eventHandleFirst.getPriority().ordinal();
            }
        });
    }

    /**
     * Removes all methods in the object from the {@link #eventHandleMap}
     * @param object the object of which you want the listeners removed from
     */
    public void removeEventListener(Object object)
    {
        Iterator<Map.Entry<Class<?extends Event>, List<EventHandle>>> iterator = eventHandleMap.entrySet().iterator();
        while(iterator.hasNext())
        {
            Map.Entry<Class<?extends Event>, List<EventHandle>> entry = iterator.next();
            Iterator<EventHandle> eventHandleIterator = entry.getValue().iterator();
            while(eventHandleIterator.hasNext())
            {
                if(eventHandleIterator.next().getMethodClass().equals(object))
                    eventHandleIterator.remove();
            }
        }
    }

    /**
     * Fire an events to all applicable events handles
     * @param event the events to be fired
     */
    public void fireEvent(Event event)
    {
        List<EventHandle> eventHandles;
        if((eventHandles = eventHandleMap.get(event.getClass())) != null)
        {
            for(EventHandle eventHandle : eventHandles)
            {
                if(!event.isCancelled() && eventHandle.getPriority() != EventPriority.MONITOR) {
                    try {
                        eventHandle.getMethod().invoke(eventHandle.getMethodClass(), event);
                    } catch (InvocationTargetException | IllegalAccessException e) {
                        e.printStackTrace();
                    }
                }
            }
        }
    }
}

File: src/main/java/dev/throwouterror/eventbus/SimpleEventManager.kt
package dev.throwouterror.eventbus

import dev.throwouterror.eventbus.annotation.EventHandler
import dev.throwouterror.eventbus.annotation.EventPriority
import dev.throwouterror.eventbus.events.Event
import java.lang.reflect.InvocationTargetException
import java.util.*
import kotlin.collections.ArrayList

class SimpleEventManager : IEventManager {
    private val eventHandleMap: MutableMap<Class<out Event>, MutableList<EventHandle>> = HashMap()

    /**
     * Add an object as an events listener, this method will loop through the objects methods
     * and if the method has [dev.throwouterror.eventbus.annotation.EventHandler] present then it will add it has an Event Handler
     * @param object the object to register
     */
    override fun addEventListener(`object`: Any) {
        for (method in `object`.javaClass.declaredMethods) if (method.isAnnotationPresent(EventHandler::class.java) && method.parameterTypes.isNotEmpty()) {
            val eventHandler: EventHandler = method.declaredAnnotations[0] as EventHandler
            if (eventHandleMap[method.parameterTypes[0]] != null) {
                eventHandleMap[method.parameterTypes[0]]!!.add(EventHandle(method, `object`, eventHandler.priority))
                sortArrayList(eventHandleMap[method.parameterTypes[0]]!!)
            } else {
                val eventHandles: MutableList<EventHandle> = ArrayList()
                eventHandles.add(EventHandle(method, `object`, eventHandler.priority))
                eventHandleMap[method.parameterTypes[0] as Class<out Event>] = eventHandles
            }
            method.isAccessible = true
        }
    }

    /**
     * Add an object as an events listener, this method will loop through the objects methods
     * and if the method has [dev.throwouterror.eventbus.annotation.EventHandler] present then it will add it has an Event Handler
     * @param object the object to register
     * @param eventClass the specific events class we want to filter for
     */
    override fun addSpecificEventListener(`object`: Any, eventClass: Class<out Event?>) {
        for (method in `object`.javaClass.declaredMethods) if (method.isAnnotationPresent(EventHandler::class.java) && method.parameterTypes.isNotEmpty() && method.parameterTypes[0] == eventClass) {
            val eventHandler: EventHandler = method.declaredAnnotations[0] as EventHandler
            if (eventHandleMap[method.parameterTypes[0]] != null) {
                eventHandleMap[method.parameterTypes[0]]!!.add(EventHandle(method, `object`, eventHandler.priority))
                sortArrayList(eventHandleMap[method.parameterTypes[0]]!!)
            } else {
                val eventHandles: MutableList<EventHandle> = ArrayList()
                eventHandles.add(EventHandle(method, `object`, eventHandler.priority))
                eventHandleMap[method.parameterTypes[0] as Class<out Event>] = eventHandles
            }
            method.isAccessible = true
        }
    }

    /**
     * Sort the ArrayList by the Priority making sure its Highest - Lowest
     * @param arrayList the ArrayList to sort
     */
    private fun sortArrayList(arrayList: MutableList<EventHandle>) {
        arrayList.sortWith(Comparator { objOne, objTwo ->
            objTwo!!.getPriority().ordinal - objOne!!.getPriority().ordinal
        })
    }

    /**
     * Removes all methods in the object from the [.eventHandleMap]
     * @param object the object of which you want the listeners removed from
     */
    override fun removeEventListener(`object`: Any?) {
        val iterator: Iterator<Map.Entry<Class<out Event>, MutableList<EventHandle>>> = eventHandleMap.entries.iterator()
        while (iterator.hasNext()) {
            val entry: Map.Entry<Class<out Event>, MutableList<EventHandle>> = iterator.next()
            val eventHandleIterator: MutableIterator<EventHandle> = entry.value.iterator()
            while (eventHandleIterator.hasNext()) {
                if (eventHandleIterator.next().methodClass == `object`) eventHandleIterator.remove()
            }
        }
    }

    /**
     * Fire an events to all applicable events handles
     * @param event the events to be fired
     */
    override fun fireEvent(event: Event) {
        val eventHandles = getHandlers(event::class.java)
        for (eventHandle in eventHandles) {
            if (!event.isCancelled && eventHandle.getPriority() !== EventPriority.MONITOR) {
                try {
                    eventHandle.method.invoke(eventHandle.methodClass, event)
                } catch (e: InvocationTargetException) {
                    e.printStackTrace()
                } catch (e: IllegalAccessException) {
                    e.printStackTrace()
                }
            }
        }
    }

    fun getHandlers(event: Class<out Event>): MutableList<EventHandle> {
        return if (eventHandleMap[event] != null) eventHandleMap[event]!! else arrayListOf()
    }
}